app/controllers/openid_connect/authorization_controller.rb

```diff
@@ -94,7 +94,7 @@ def email_address_id
         return user_session[:selected_email_id_for_linked_identity]
       end
       identity = current_user.identities.find_by(service_provider: sp_session[:issuer])
-      identity&.email_address_id
+      identity&.email_address_for_sharing&.id
```

Comment:
I'm not sure I follow these changes. Since `email_address_for_sharing` falls back to the `last_sign_in_email_address` (instead of `nil`), won't this cause us to start linking email addresses where we don't want to be linking them?

Also, with any changes to OIDC behavior, I'd expect to see equivalent changes to the SAML implementation as well.